Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs

```diff
@@ -0,0 +1,319 @@
+
+                var totalSlots = 0;
+
+                foreach (var (_, slot) in stationTotalSlots)
+                {
+                    totalSlots += slot;
+                }
+
+                if (totalSlots == 0)
+                    continue; // No slots so just leave.
+
+                // Clear for reuse.
+                stationShares.Clear();
+                var distributed = 0;
+
+                foreach (var station in stations)
+                {
+                    // Calculates the percent share then multiplies.
+                    stationShares[station] = (int)Math.Floor(((float)stationTotalSlots[station] / totalSlots) * candidates.Count);
+                    distributed += stationShares[station];
+                }
+
+                // Avoids the fair share problem where if there's two stations and one player neither gets one.
+                if (distributed < candidates.Count)
+                {
+                    var choice = _random.Pick(stations);
+                    stationShares[choice] += candidates.Count - distributed;
+                }
+
+                // Actual meat, goes through each station and shakes the tree until everyone has a job.
+                foreach (var station in stations)
+                {
+                    if (stationShares[station] == 0)
+                        continue;
+
+                    var slots = stationSlots[station];
+                    var allJobs = slots.Keys.ToList();
+                    _random.Shuffle(allJobs);
+                    // And iterates through all it's jobs in a random order until the count settles.
+                    // No, AFAIK it cannot be done any saner than this. I hate "shaking" collections as much
+                    // as you do but it's what seems to be the absolute best option here.
+                    // It doesn't seem to show up on the chart, perf-wise, anyway, so it's likely fine.
+                    int priorCount;
+                    do
+                    {
+                        priorCount = stationShares[station];
+
+                        foreach (var job in allJobs)
+                        {
+                            if (stationShares[station] == 0)
+                                break;
+
+                            if (slots[job] != null && slots[job] == 0)
+                                continue; // Can't assign this job.
+
+                            if (!jobPlayerOptions.ContainsKey(job))
+                                continue;
+
+                            // Picking players it finds that have the job set.
+                            var player = _random.Pick(jobPlayerOptions[job]);
+                            AssignPlayer(player, job, station);
+                            stationShares[station]--;
+
+                            if (slots[job] != null)
+                                slots[job]--;
+
+                            if (optionsRemaining == 0)
+                                goto done;
```

Comment:
death

Reply:
can this not be done by just storing a failure bool

Reply:
I think this is like the one valid use for goto, tbh.

Reply:
you'd have to check the failure bool constantly.
`goto label` is fine tbh.